Plugin dev quickstart (#7068)
Kong/docs.konghq.com: app/_src/gateway/plugin-development/get-started/testing.md

### 2. Initialize the test environment

Pongo allows us to validate a plugin's behavior by giving us tools to quickly run a 
{{site.base_gateway}} with our plugin installed and available. Let's validate the plugin 
manually first then we will add automated tests in subsequent steps of this guide. 

{:.note}
> **Note**: {{site.base_gateway}} runs in a variety of
> [deployment topologies](/gateway{{page.release}}/production/deployment-topologies). 
> By default, Pongo runs {{site.base_gateway}} in _traditional mode_ which uses a database 
> to store configured entities such as routes, services, and plugins. 
> {{site.base_gateway}} and the database are ran in separate containers allowing us
> to cycle the gateway independently of the database. This enables a quick
> iterative approach to validating the plugin's logical behavior while keeping the gateway
> state independent in the database.

Pongo provides an optional command that will initialize the project directory with some
default configuration files which can be ran one time when starting a new project.

{:.important}
> **Important:** These commands must be ran from the base `my-plugin` directory so Pongo properly 

> > packages and includes the plugin code in the running {{site.base_gateway}}.

Author: Yes, this is important.  I hope my use of "base" clarifies, but i'm open to other suggestions.

Initialize the project folder:

```sh
pongo init
```

Now we are ready to start dependency containers for {{site.base_gateway}}. By default
this only includes the Postgres database used in traditional mode.

Start the dependencies: 

```sh

> pongo up

Author: I tried adding numbered lists and was not happy with how it looked due to some sections with only one or two instructions it also required indenting significant sections of copy.  I'm not really convinced it helps the user to have many small sub numbered lists within the guides.

requests via the `example_service`. Prior to forwarding the response from the 
upstream, our plugin should append the `X-MyPlugin` header to the list of response headers.

Let's test the behavior by proxying a request and asking `curl` to show the 
response headers with the `-i` flag:

```sh
curl -i http://localhost:8000/mock/anything
```

`curl` should report `HTTP/1.1 200 OK` and show us the response headers from the gateway. Included
in the set of headers, should be `X-MyPlugin: response`, indicating that our plugin logic has been invoked.

For example: 

```sh
HTTP/1.1 200 OK
Content-Type: application/json
Connection: keep-alive
Content-Length: 529
Access-Control-Allow-Credentials: true
Date: Tue, 12 Mar 2024 14:44:22 GMT
Access-Control-Allow-Origin: *
Server: gunicorn/19.9.0
X-MyPlugin: response
X-Kong-Upstream-Latency: 97
X-Kong-Proxy-Latency: 1
Via: kong/3.6.1
X-Kong-Request-Id: 8ab8c32c4782536592994514b6dadf55
```

Exit the {{site.base_gateway}} shell before proceeding:

```sh
exit
```

For quickly getting started, manually validating a plugin using the Pongo shell works
nicely. However, you will prefer to deploy automated testing and maybe a Test-driven development 

> (TDD) methodology. Let's see how Pongo can help with this as well.

Author: @lena-larionova I reworked this section.  I didn't like the "real-life" bit, but tried to capture your concern with "production scenarios"

### 4. Write a test

Pongo supports running automated tests using the 
[Busted](https://lunarmodules.github.io/busted/) Lua test framework. In plugin

> projects, the test files reside under the `spec/<plugin-name>` directory. For this project
> we created the `spec/my-plugin` folder earlier.  

Reviewer: ```suggestion
projects, the test files reside under the `spec/<plugin-name>` directory. 
For this project, this is the `spec/my-plugin` folder you created earlier.  
```

> The following is a code listing for a test that validates our plugin's current behavior. 

Reviewer: ```suggestion
The following is a code listing for a test that validates the plugin's current behavior. 
```

> Copy this code and place it into a new file located at `spec/my-plugin/01-integration_spec.lua`. 

Reviewer: For parallelism, there should be a `touch` command here first:

```
touch spec/my-plugin/01-integration_spec.lua
```
(can't add as GH suggestion due to the backticks)

Author: I don't agree with this one.  The reason for the `touch` commands in the setup section was that I wanted to teach the user about required modules and the "setup" of a new project. Having them touch the files first helped show required files.  I don't think having the user `touch` files here and then open an editor and paste is necessary here.

        nginx_conf = "spec/fixtures/custom_nginx.template",
        -- make sure our plugin gets loaded
        plugins = "bundled," .. PLUGIN_NAME,
      }))

    end)

    -- teardown runs after its parent describe block
    teardown(function()
      helpers.stop_kong(nil, true)
    end)

    -- before_each runs before each child describe
    before_each(function()
      client = helpers.proxy_client()
    end)

    -- after_each runs after each child describe
    after_each(function()
      if client then client:close() end
    end)

    -- a nested describe defines an actual test on the plugin behavior
    describe("The response", function()

      it("gets the expected header", function()

        -- invoke a test request
        local r = client:get("/mock/anything", {})

        -- validate that the request succeeded, response status 200
        assert.response(r).has.status(200)

        -- now validate and retrieve the expected response header 
        local header_value = assert.response(r).has.header("X-MyPlugin")

        -- validate the value of that header
        assert.equal("response", header_value)

      end)
    end)
  end)
end
```

> We now have test code, let's see how Pongo can help us automate testing.

Reviewer: ```suggestion
With this test code, Pongo can help automate testing.
```

> ### 5. Run the test

Reviewer: ```suggestion
### Run the test
```

> Pongo can run automated tests for us with the `pongo run` command. When this is executed,

Reviewer: ```suggestion
Pongo can run automated tests with the `pongo run` command. When this is executed,
```

> Pongo will determine if dependency containers are already running and use them
> if so. The test library will handle truncating existing data in between test runs for us.

Reviewer: ```suggestion
Pongo determines if dependency containers are already running and use them if they are. 
The test library handles truncating existing data in between test runs.
```

Author: @lena-larionova I reworked this but did not take the suggestion.

> Once the test file from the previous step is saved in the `spec/my-plugin/01-integration_spec.lua` file, 
> execute a test run:

Reviewer: ```suggestion
Execute a test run:
```
The rest of the sentence feels like overkill.

```sh
[pongo-INFO] auto-starting the test environment, use the 'pongo down' action to stop it
Kong version: 3.6.1

[==========] Running tests from scanned files.
[----------] Global test environment setup.
[----------] Running tests from /kong-plugin/spec/my-plugin/01-integration_spec.lua
[----------] Running tests from /kong-plugin/spec/my-plugin/01-integration_spec.lua
[ RUN      ] /kong-plugin/spec/my-plugin/01-integration_spec.lua:63: my-plugin: [#postgres] The response gets a 'X-MyPlugin' header
[       OK ] /kong-plugin/spec/my-plugin/01-integration_spec.lua:63: my-plugin: [#postgres] The response gets a 'X-MyPlugin' header (6.59 ms)
[ RUN      ] /kong-plugin/spec/my-plugin/01-integration_spec.lua:63: my-plugin: [#off] The response gets a 'X-MyPlugin' header
[       OK ] /kong-plugin/spec/my-plugin/01-integration_spec.lua:63: my-plugin: [#off] The response gets a 'X-MyPlugin' header (4.76 ms)
[----------] 2 tests from /kong-plugin/spec/my-plugin/01-integration_spec.lua (23022.12 ms total)

[----------] Global test environment teardown.
[==========] 2 tests from 1 test file ran. (23022.80 ms total)
[  PASSED  ] 2 tests.
```

> Pongo can also run as part of a Continuous Integration (CI) system, the 
> [repository documentation](https://github.com/Kong/kong-pongo?tab=readme-ov-file#setting-up-ci) has more details. 

Reviewer: ```suggestion
Pongo can also run as part of a Continuous Integration (CI) system. See the 
[repository documentation](https://github.com/Kong/kong-pongo?tab=readme-ov-file#setting-up-ci) for more details. 
```

```sh
pongo help
```

With Pongo installed, you can now set up a test environment for your new plugin.

### Initialize the test environment

Pongo lets you validate a plugin's behavior by giving you tools to quickly run a 
{{site.base_gateway}} with the plugin installed and available. 

Let's validate the plugin 
manually first, and then you will add automated tests in subsequent steps of this guide. 

{:.note}
> **Note**: {{site.base_gateway}} runs in a variety of
> [deployment topologies](/gateway{{page.release}}/production/deployment-topologies). 
> By default, Pongo runs {{site.base_gateway}} in _traditional mode_, which uses a database 
> to store configured entities such as routes, services, and plugins. 
> {{site.base_gateway}} and the database are run in separate containers,
> letting you cycle the gateway independently of the database. This enables a quick and 
> iterative approach to validating the plugin's logical behavior while keeping the gateway
> state independent in the database.

Pongo provides an optional command that initializes the project directory with some
default configuration files. You can run it to start a new project.

{:.important}

> > **Important:** These commands must be run from the base `my-plugin` directory so that Pongo properly 

Author: I'm fine with `inside` but there are two `my-plugin` directories. @lena-larionova pointed out it may be confusing, I was hoping the use of the term `base` conveyed the project root.  So maybe I should use "project root folder (my-plugin)" instead.

> > **Important:** These commands must be run from the base `my-plugin` directory so that Pongo properly 

Author: please see revised version of this.  Thank you for the reviews @mheap and @lena-larionova
